src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -101,6 +101,14 @@ private void CreateNodeCaches()
                     {
                         return new GenericDefinitionEETypeNode(this, type);
                     }
+                    else if (type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+                    {
+                        return new CanonicalDefinitionEETypeNode(this, type);
+                    }
+                    else if (type.IsCanonicalSubtype(CanonicalFormKind.Any))
+                    {
+                        return new CanonicalEETypeNode(this, type);
```

Comment:
The concept of necessary and constructed should not apply to canonical instantiations... Canonical types' existence only makes sense as template types, which is different.


Today, there are some parts of the code that will ask for a necessary type, and some other parts that would has for a constructed type. Regardless of who asks for a canonical instantiation type, these types will become template types at the end. They are neither necessary minimal types, nor constructed full types. They are template EETypes that will need to just contain the necessary pieces used by the dynamic type loader (ex: interface dispatch map, vtable, and for non-USG: flags and gcdesc).

I like the idea of putting the canonical ones in a separate nodecache, but I don't like the idea of explicitly asking for them using a new method on NodeFactory: the canonical instantiation type will have some functionality shared with the EETypeNode baseclass, and I don't want to introduce code in multiple places that explicitly asks for different type symbols just because a type is canonical. I prefer we just implement the differences in the derived CanonicalEETypeNode.